### DISCUSSION TITLE
ServiceStack.Text Json Performance

### DISCUSSION THREAD

**Discussion by mylvghb** [ServiceStack/Discuss#97] [Q&A] [answered]
I compared three JSON serialization and deserialization frameworks, and after testing, I found that the performance of ServiceStack was not ideal.

Code:

```csharp
var faker = new Faker<User>()
    .CustomInstantiator(f => new User(
        Guid.NewGuid(),
        f.Name.FirstName(),
        f.Name.LastName(),
        f.Name.FullName(),
        f.Internet.UserName(f.Name.FirstName(), f.Name.LastName()),
        f.Internet.Email(f.Name.FirstName(), f.Name.LastName())
    ));

Stopwatch stopwatch;
string text;
var users  = faker.Generate(10000);

stopwatch = Stopwatch.StartNew();
text = Newtonsoft.Json.JsonConvert.SerializeObject(users);
stopwatch.Stop();
Console.WriteLine($"JSON.NET Serialized {users.Count} users in {stopwatch.ElapsedMilliseconds}ms");

stopwatch = Stopwatch.StartNew();
text = JsonSerializer.SerializeToString(users);
stopwatch.Stop();
Console.WriteLine($"ServiceStack Serialized {users.Count} users in {stopwatch.ElapsedMilliseconds}ms");

stopwatch = Stopwatch.StartNew();
text = System.Text.Json.JsonSerializer.Serialize(users);
stopwatch.Stop();
Console.WriteLine($"Microsoft Serialized {users.Count} users in {stopwatch.ElapsedMilliseconds}ms");
```

Results:

JSON.NET Serialized 10000 users in 117ms
ServiceStack Serialized 10000 users in 195ms
Microsoft Serialized 10000 users in 42ms

Can the performance of ServiceStack in JSON processing be improved?

**Comment by mylvghb** [comment]
Now, Microsoft's JSON serializer stands out in terms of performance.

**Reply by mythz** [reply]
Can you put this in a new project so we can clone and run the same implementation please.

**Reply by mylvghb** [reply]
Okay

**Reply by mylvghb** [reply]
I put it on github: https://github.com/mylvghb/JsonPerformanceTests.git

**Reply by mylvghb** [reply]
Wait a minute, there's something wrong

**Reply by mylvghb** [reply]
All codes have been submitted

**Comment by mythz** [comment]
I'm assuming you got an unpredictable run as the first 3 runs after build shows ServiceStack.Text is consistently between JSON.NET and `System.Text.Json` which is expected:

<img width="665" alt="image" src="https://github.com/ServiceStack/Discuss/assets/89361/57c2f316-0ec6-4c5d-9e63-b98fb6314c57">

Going to ignore unless I can repro it being consistently slower than JSON.NET.

**Reply by mylvghb** [reply]
On my computer are:  
![image](https://github.com/ServiceStack/Discuss/assets/35945793/d5454f1b-8e8d-40a2-bdc7-a085a9857d2a)

**Reply by mylvghb** [reply]
![image](https://github.com/ServiceStack/Discuss/assets/35945793/9fe30950-dcdb-4a7f-92bd-586c5e2f2a85)

**Reply by mythz** [reply]
What OS/arch are you running this on?

**Reply by mylvghb** [reply]
Win10

**Reply by mylvghb** [reply]
![image](https://github.com/ServiceStack/Discuss/assets/35945793/40eab1cb-b3ef-4f43-8e75-554e3cff8f79)

**Reply by mythz** [reply]
I'm also using Win 10. Can you try running a production build to better test real world performance:

```sh
$ dotnet publish
$ cd .\bin\Release\net8.0\publish\
$  dotnet .\JsonPerformanceTests.dll
```

This is also consistently between JSON.NET and System.Text.Json

<img width="881" alt="image" src="https://github.com/ServiceStack/Discuss/assets/89361/14b0db18-fc57-4432-8076-7ad327d69ba5">

**Reply by mylvghb** [reply]
![image](https://github.com/ServiceStack/Discuss/assets/35945793/48f6aec1-e04c-45b3-93ed-ca9a3bee35d1)

**Reply by mylvghb** [reply]
Could you please run the benchmark?
```csharp
var summary = BenchmarkRunner.Run<SerializeBigDataBenchmark>();
Console.WriteLine(summary);
```

**Reply by mythz** [reply]
Can't explain why you're seeing different results, getting similar results on my M2 MacBook Air:

<img width="944" alt="image" src="https://github.com/ServiceStack/Discuss/assets/89361/e35e84a8-9d8b-447f-a21f-37329bd77bce">

**Reply by mythz** [reply]
Ok yeah, this shows being marginally slower than JSON.NET

<img width="938" alt="image" src="https://github.com/ServiceStack/Discuss/assets/89361/e4c8c8f0-6719-4b27-9f60-9f998308bf26">

But not enough to prioritize over other features we're working on, may revisit later.

**Reply by mylvghb** [reply]
Other important features take priority, as long as you pay attention to this issue.

**Reply by mylvghb** [reply]
I'm just raising the issues I found

**Comment by mythz** [comment, marked as answer]
FYI from v8.1 ServiceStack now [uses System.Text.Json by default](https://docs.servicestack.net/releases/v8_01#system.text.json) in its Endpoint Routing APIs for Serializing/Deserializing its JSON APIs.

**Comment by mylvghb** [comment]
Thank you for your timely notification!
